### ISSUE TITLE
sql/catalog: cannot set a descriptor name to a name being queried

### ISSUE THREAD

**Issue by ajwerner** [cockroachdb/cockroach#96840]
**Describe the problem**

The leasing subsystem uses high priority transactions to avoid deadlocks in some cases. Resolving descriptors uses the leasing infrastructure. The leasing infrastructure has no mechanism to cache the non-existence of a name. Every attempt to resolve a descriptor that does not exist leads to a high priority transaction on the name key; these transactions can starve normal transactions.

**To Reproduce**

Using a geo-replicated deployment makes the problem described here more prominent. This is because the starvation occurs because the transaction cannot refresh its reads before it's pushed again; a geo-replicated cluster makes it easy to make the refresh slow.

* Run a workload that queries a table which does not exist. 
* Create a table with a different name.
* Attempt to rename the table to the table name being queried by the workload.

**Proposed solution**

The reason the high priority transaction is used to avoid deadlocks. The leasing transaction is not a user transaction, but blocks progress for user transactions trying to acquire the lease. There's a singleflight. If the leasing transaction were to lock on an intent of a transaction blocked waiting on it, there'd be an undetected deadlock. In fact, this is very much a thing that happens if you run schema changes at HIGH PRIORITY. One probably shouldn't do that today (bug).

If we could only connect the waiting transactions to the transaction, we could detect this deadlock regardless of priority levels and we could let normal concurrency control do its thing.

An awkward part about this proposal is that we have to expose a way to express this dependency from the kv layer -- that always felt a bit layer violating. As a way to try to paper over some of that layer violating, this issue proposes introducing a kvclient-level concept as a library: `txnsingleflight`. 

The core idea is to expose the ability to run a transaction on behalf of N goroutines by ensuring that the kv deadlock detection mechanism will discover deadlocks between the caller txns and the singleflight txn.

The `txnsingleflight.Group` will work like the `singleflight` but it'll also asynchronously launch goroutine[s] to send a `PushTxnRequest` to the running transaction in the callback somehow.

Here's a strawman API:
```
package txnsingleflight

type Group struct {/*...*/}

func (Group) Do(
    ctx context.Context,
    key string,
    txnBlockedOnAcquire *kv.Txn,
    fn func(context.Context, *kv.Txn) (interface{}, error),
 ) (v interface{}, err error, shared bool)
```

Perhaps there will also need to be a callback that can be run in the singleflight after the transaction commits.

**Alternatives considered**
If we cached the complete set of children of descriptors in some meaningful way, we could effectively eliminate the cause of the starvation; we'd effectively have a form of negative cache. This relates in that way to https://github.com/cockroachdb/cockroach/issues/96181. These two pieces of work are complements to each other. We'd still have a space for deadlocks if we cached:

* the entire set of children for a schema, and
* the set of schemas for a database (this one we actually do cache in the database descriptor), and
* the set of database descriptors

We ought to do that anyway as it might allow us to serve virtual tables from caches.

**Expected behavior**
The rename should succeed and then the queries will start seeing the table.

**Additional context**
This issue tracks the particularly pernicious problem of not being able to plop a descriptor into a name that traffic is already interacting with. https://github.com/cockroachdb/cockroach/issues/86324 talks about a related problem for dropping descriptors. The root cause occurs for the same fundamental reason: high priority leasing transactions pushing the user. 

Jira issue: CRDB-24356

**Comment by ajwerner** [comment]
The problem with the proposal as written is that it assumes that we'll switch to using regular priority transactions to perform the read. This is hazardous in the case where a long-running schema change transaction is trying to perform some renames; the lookups will block on that rename, potentially forever. If the name did exist, this would be bad.

There is a bright future coming. The combination of https://github.com/cockroachdb/cockroach/issues/95225
and https://github.com/cockroachdb/cockroach/issues/95227.

An approach to solving this in general is to use `PRIORITY HIGH` transactions for schema changes, but those are problematic because they expose the very deadlocks we eliminated in the common case by using high priority transactions inside the leasing layer. 

Fortunately there's a clear solution for that problem on the horizon too:
* https://github.com/nvanbenschoten/cockroach/commit/9190180edebcbd782b49394ec6129b7f80a28342 **kv: push timestamp does not wait**
PUSH_TIMESTAMP does not block on PENDING pushee transactions with same
normalized priority.

We're not intending to enable the last commit for serializable transactions by default in 23.2, but we could make it available to lower-level transaction APIs.

----

That leaves us with the question of what to do in the short-term, because we should do something. The revised proposal is to introduce a knob and a small negative cache of names which we failed to resolve in the lease manager, and we do some exponential backoff before attempting to resolve those names. Hopefully that gives the restore or rename enough time to do its thing. There are costs to such negative caching, so we'll put them behind a knob and hope that somebody is paying enough attention to find it.
